kotlinx-coroutines-core/common/src/flow/operators/Errors.kt

```diff
@@ -13,90 +13,210 @@ import kotlin.coroutines.*
-                    } catch (e: Throwable) {
-                        fromDownstream = true
-                        throw e
-                    }
+    return retryWhen { cause, attempt -> predicate(cause) && attempt < retries }
+}
+
+/**
+ * Retries collection of the given flow when an exception occurs in the upstream flow and the
+ * [predicate] returns true. The predicate also receives an `attempt` number as parameter,
+ * starting from zero on the initial call. This operator is *transparent* to exceptions that occur
+ * in downstream flow and does not retry on exceptions that are thrown to cancel the flow.
+ *
+ * For example, the following call retries the flow forever if the error is caused by `IOException`, but
+ * stops after 3 retries on any other exception:
+ *
+ * ```
+ * flow.retryWhen { cause, attempt -> cause is IOException || attempt < 3 }
+ * ```
+ *
+ * To implement a simple retry logic with a limit on the number of retries use [retry] operator.
+ *
+ * Similarly to [catch] operator, the [predicate] code has [FlowCollector] as a receiver and can
+ * [emit][FlowCollector.emit] values downstream.
+ * The [predicate] is a suspending function, so it can be used to introduce delay before retry, for example:
+ *
+ * ```
+ * flow.retryWhen { cause, attempt ->
+ *     if (cause is IOException) {    // retry on IOException
+ *         emit(RetryWrapperValue(e))
+ *         delay(1000)                // delay for one second before retry
+ *         true
+ *     } else {                       // do not retry otherwise
+ *         false
+ *     }
+ * }
+ * ```
+ *
+ * See [catch] for more details.
+ */
+@ExperimentalCoroutinesApi // tentatively stable in 1.3.0
+public fun <T> Flow<T>.retryWhen(predicate: suspend FlowCollector<T>.(cause: Throwable, attempt: Long) -> Boolean): Flow<T> =
+    flow {
+        var attempt = 0L
+        var shallRetry: Boolean
+        do {
+            shallRetry = false
+            catchImpl(this) { e ->
+                if (predicate(e, attempt)) {
+                    shallRetry = true
+                    attempt++
+                } else {
+                    throw e
                 }
-                break
+            }
+        } while (shallRetry)
+    }
+
+// Note that exception may come from the downstream operators, we should not catch them
+private suspend inline fun <T> Flow<T>.catchImpl(
+    collector: FlowCollector<T>,
+    action: FlowCollector<T>.(Throwable) -> Unit
+) {
+    var fromDownstream: Throwable? = null
+    try {
+        collect {
+            try {
+                collector.emit(it)
             } catch (e: Throwable) {
-                if (fromDownstream || e.isCancellationCause(coroutineContext)) throw e
-                if (!predicate(e) || retries-- == 0) throw e
+                fromDownstream = e
+                throw e
             }
         }
+    } catch (e: Throwable) {
+        if (e.isSameExceptionAs(fromDownstream) || e.isCancellationCause(coroutineContext))
```

Comment:
Is my understanding correct that you explicitly check for cancellation exception as well because there's no point in allowing the operator to see an exception if it has already caused the coroutine to be cancelled, since the operator would effectively be a no-op in that case anyway?

Reply:
>since the operator would effectively be a no-op in that case anyway?

It won't be a no-op and that it the problem.
Consider the following code:
```
flow {
   // suspension point (1)
}
.catch({ println(it)  }) // Print an exception
.map {
  // suspension point (2)
}
.single()
```
If the enclosing coroutine is cancelled, `CancellationException` will be thrown either from `(1)` or `(2)`. `catch` (and `retry`) **should** be consistent in both cases.